File: src/Java/algorithms/arrandstring/MergeSortedArray.java
package algorithms.arrandstring; 

import java.util.*; 

public class MergeSortedArray {
    /* Function: A and B are sorted arrays in first m and n cells. Now merge B into A. 
     * Solution: Three pointers. 
     */ 
    
    public void merge(int A[], int m, int B[], int n) {
        if (A == null || B == null || B.length < n || A.length < m + n) {
            return; 
        }
        
        int pt = m + n - 1, pt1 = m - 1, pt2 = n - 1; 
        while (pt1 >= 0 && pt2 >= 0) {
            if (A[pt1] > B[pt2]) {
                A[pt--] = A[pt1--]; 
            } else {
                A[pt--] = B[pt2--]; 
            }
        }
        
        while (pt2 >= 0) {
            A[pt--] = B[pt2--]; 
        }
    }
} 

File: src/CPP/src/leetcode/MergeSortedArray.cpp
/*
 * [Source] https://leetcode.com/problems/merge-sorted-array/
 * [Difficulty]: Easy
 * [Tag]: Array
 * [Tag]: Two Pointers
 */

#include <iostream>

using namespace std;

// [Solution]: Start from end
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/arrandstring/MergeSortedArray.java
 */

/* Java solution
public class Solution {
    public void merge(int A[], int m, int B[], int n) {
		if(A.length<m+n) return;
        if(n==0) return;
		if(m==0){
			for(int i=0; i<n; i++){
				A[i]=B[i];
			}
			return;
		}
		int ae=m-1;
		int be=n-1;
		int end=m+n-1;
		while(ae>=0 && be>=0){
			if(A[ae]>B[be]) {
				A[end]=A[ae];
				ae--;
			}else{
				A[end]=B[be];
				be--;
			}
			end--;
		}
		if(be<0) return;
		while(be>=0){
			A[end]=B[be];
			be--;
			end--;
		}
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
